### ISSUE TITLE
Mitigating Cross-Protocol Attacks

### ISSUE THREAD

**Issue by thejh** [WICG/cors-rfc1918#8]
From Twitter thread: https://twitter.com/tehjh/status/788369142550192128

There are protocols (well, at least one) in which it is possible to get cross-protocol RCE just using the request path: https://blog.lizzie.io/exploiting-CVE-2016-8606.html

Because a CORS request still contains the request path, this proposal wouldn't mitigate that attack.

IMO it would make sense to add a rule that triggers on the following conditions:

 - this is the first request on the socket (for performance, could be omitted)
 - the destination port is not a typical HTTP port (for performance)
 - the protocol is HTTP

and, when all conditions are fulfilled, first sends an "are you an HTTP server?" `OPTIONS` preflight or so without *any* user-controlled data (no `Host` or `Host` derived from IP address, empty request path, ...) and checks whether the response looks like HTTP.

Many protocols use newlines or spaces as separators, and it might not be all that common for a protocol to be exploitable without those. But still, I think that for a robust defense, this protaection should be added. (Of course, that's easy for me to say, considering that I won't have to implement all this.)

**Comment by codedokode** [comment]
I would like to bump this issue and give another example of vulnerability possible even with this proposal. 

Some web servers are configured to ignore HTTP method and run a script based on path only. For example, Apache with PHP will run script `execute_shell_command.php` in response to request with any method, including `OPTIONS /execute_shell_command.php?cmd=rm+/+-rf`. Nginx docs contain [example configuration](https://www.nginx.com/resources/wiki/start/topics/examples/phpfcgi/#connecting-nginx-to-php-fpm) that will also route all requests to a specific path to a script. Linux distributions like Debian have Apache or Nginx configured this way by default. So we can assume that in most cases if a vulnerable device on a private network uses PHP then this proposal won't protect it from executing request from a public Internet.

Of course, this is not limited to PHP. A web-server written in Go or Rust could also check only request path and ignore HTTP method because it requires less code.

As one more example, ASP.NET server also [supports routing based on path only](https://docs.microsoft.com/en-us/previous-versions/cc668201(v=vs.140)?redirectedfrom=MSDN): 

> For example, a request for https://server/application/Products.aspx?id=4 maps to a file that is named Products.aspx that contains code and markup for rendering a response to the browser.

To mitigate this I suggest that browser would send a preflight request to root URL and specify requested path in the headers, for example:

```http
OPTIONS / HTTP/1.1
Requested-path: /execute_shell_command.php?cmd=rm+/+-rf
...
```
Sending request to `/` is safe in most cases.

**Comment by letitz** [comment]
That's true, thanks for pointing it out!

Behaving like you suggest would deviate even further from CORS, requiring more invasive changes to web servers' CORS code as well as browsers (but that's less important). I'm not sure the complexity warrants the security benefit - especially since at that point the server is *very* broken.

CORS already includes the path in the request and seems relatively successful in mitigating CSRF attacks on the web at large, which leads me to believe it is good enough?

**Comment by mikewest** [comment]
I think it's reasonable to consider this kind of change for CORS generally (though I think it would be quite difficult to ship). I'm less sure it's reasonable to tie Private Network Access to it, as the threat to internal and external servers is equivalent.

**Comment by letitz** [comment]
Fair point. Let's move this discussion to the [fetch repository](https://github.com/whatwg/fetch) (CORS is defined [in Fetch](https://fetch.spec.whatwg.org/#http-cors-protocol)) then?

**Comment by codedokode** [comment]
Ok, indeed this is better to discuss in Fetch repository. Of course, it would be better if path was removed from URL for all preflight requests, not only for those targeting private IP addresses.

>  I'm not sure the complexity warrants the security benefit - especially since at that point the server is very broken.

Actually I would like even more changes. It would be better if starting with next HTTP version (for example, HTTP/3 or HTTP/4) *all* cross-origin requests (including sending a form or fetching an image) would be implemented as opt-in (the receiving server should explicitly agree to receive them). Currently a developer is required to make some extra effort to protect application from CSRF (for example: check request method, check `Origin` header, check CSRF tokens) and there always will be lazy developers who won't do this. Making cross-origin requests *opt-in* and "secure by default" would finally solve this problem that was created many years ago.

**Comment by annevk** [comment]
It's an interesting idea, but we cannot really tie those kind of changes to new iterations of the HTTP protocol. That would hamper adoption too much. Instead we'll continue to iterate to make those kind of requests less impactful across the board.
